packages/formula-editor/src/formula-editor.ts

```diff
@@ -0,0 +1,219 @@
+
+    const parseOutput = this._parser.parseInput(
+      this._content,
+      this.currentCursorPosition,
+      addRecommendation
+    );
+
+    this._recommendations = parseOutput.recommendations;
+    this._formattedContent = parseOutput.formattedContent;
+    this._errorStr = parseOutput.errorStr;
+
+    /**  
+     * Don't modify the text stream manually if the text is being composed,
+     * unless the user manually chooses to do so by selecting a suggestion.
+     * @see https://github.com/w3c/input-events/issues/86
+     * @see https://github.com/w3c/input-events/pull/122
+     * @see https://bugs.chromium.org/p/chromium/issues/detail?id=689541
+     * */
+
+    if (this.lastInputType != "insertCompositionText" || addRecommendation) {
+      editor.innerHTML = parseOutput.formattedString!;
+    }
+
+    this._content = (editor as HTMLDivElement).innerText;
+
+    if (addRecommendation) {
+      this._recommendations = null;
+      this.currentCursorPosition = parseOutput.newCursorPosition;
+    }
+
+    Cursor.setCaret(this.currentCursorPosition!, editor);
+    editor?.focus();
+
+    this.currentCursorRect = Cursor.getCursorRect();
+    this.requestUpdate();
+  }
+
+  requestCalculate() {
+    if (this._parser.parseInput(this._content).errorStr) {
+      return;
+    }
+
+    const calculatedResult = this._parser.calculate(this._content);
+
+    this._content = this._parser.addParentheses(this._content) ?? this._content;
+    this.parseInput();
+
+    this._calculatedResult = calculatedResult ?? NaN;
+    this._errorStr =
+      calculatedResult == undefined
+        ? "Division by zero encountered"
+        : this._errorStr;
+
+    this._recommendations = null;
+    this.requestUpdate();
+  }
+
+  requestFormat() {
+    this._content = this._parser.addParentheses(this._content) ?? this._content;
+    this.parseInput();
+    this._recommendations = null;
+    this.requestUpdate();
+  }
+
+  // Disable shadow-root as it messes up cursor detection.
+  createRenderRoot() {
+    return this;
+  }
+
+  render() {
+    return html`
+      <style>
+        ${FormulaEditorStyles}
+        ${TextButtonStyles}
+      </style>
+      <div
+        contenteditable
+        id="wysiwyg-editor"
+        spellcheck="false"
+        autocomplete="off"
+        @input=${this.handleChange}
+        @keydown=${this.handleTab}
+      ></div>
+      ${this._recommendations
+        ? html` <suggestion-menu
+            style="
+              position: absolute; 
+              left: ${this.currentCursorRect?.left + "px"}; 
+              top: ${(this.currentCursorRect?.top ?? 0) +
+            window.scrollY +
+            "px"};
+            "
+            .recommendations=${this._recommendations}
+            .onClickRecommendation=${(e: any) => this.onClickRecommendation(e)}
+          ></suggestion-menu>`
+        : html``}
+      <div id="wysiwyg-err" class="${this._errorStr ?? "wysiwyg-no-err"}">
+        ${this._errorStr ?? "No Errors"}
```

Comment:
can choose to show \n instead of "No Errors"

Reply:
Newlines are treated the same as spaces, so I am changing to no breaking space instead.